src/Sylius/Bundle/ShippingBundle/Form/EventListener/AddCodeFormSubscriber.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\ShippingBundle\Form\EventListener;
+
+use Sylius\Component\Resource\Model\CodeAwareInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\Intl\Exception\UnexpectedTypeException;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+class AddCodeFormSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @return array
+     */
+    public static function getSubscribedEvents()
+    {
+        return array(
+            FormEvents::PRE_SET_DATA => 'preSetData',
+        );
+    }
+
+    /**
+     * @param FormEvent $event
+     */
+    public function preSetData(FormEvent $event)
+    {
+        $entityWithImmutableCode = $event->getData();
+        $form = $event->getForm();
+
+        if (!$entityWithImmutableCode instanceof CodeAwareInterface) {
+            throw new UnexpectedTypeException('Expected type was: ', CodeAwareInterface::class);
```

Comment:
IIRC, this is not how this exception works, you just pass the $resource variable as first value and second argument is expected type. It will generate nice exception message automatically.